libminifi/include/utils/file/FileView.h

```diff
@@ -0,0 +1,184 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <filesystem>
+#include <fstream>
+#include <string>
+#include <memory>
+
+#include "utils/gsl.h"
+#include "core/logging/LoggerConfiguration.h"
+
+#ifdef WIN32
+#include <stdio.h>
+#include <Windows.h>
+#else
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#endif
+
+namespace org::apache::nifi::minifi::utils::file {
+
+class FileView {
+  static std::string getLastError() {
+#ifdef WIN32
+    return std::system_category().message(GetLastError());
+#else
+    return std::system_category().message(errno);
+#endif
+  }
+#ifdef WIN32
+  struct FileHandle {
```

Comment:
In this file, there is a lot of unnecessary complexity just to search in a memory-mapped file instead of using fstream to do conventional read. If there is another reason for this, then could you explain it? If not, could you convert the verification code to normal buffered file io code and remove these mmap wrappers?

Reply:
`std::ifstream` is an input iterator and not a forward iterator, so `std::search` does not operate on it, we could read the file chunk by chunk but then we would have to cover edge cases like the id appearing on the boundary, tried wrapping the `std::ifstream` to accommodate it for `std::search` but it took tens of seconds with the constant seeks, the memory-mapped solution runs in tens of milliseconds, also we could use the Boyer-Moore searcher

I think it is best to use standard library utilities whenever possible instead of hand-rolling our own search implementation

Reply:
In general I agree about using the standard library, but here the cost is about 150 lines of code. I want to avoid that bloat if possible, even if that takes a few extra lines on the search side.

Reply:
To be clear, my objection is about code bloat and more maintenance burden, not memory-mapped io itself. If you want to use memory mapped io, having a third party platform-independent lib doing that would be an improvement IMO. [mio](https://github.com/mandreyel/mio) looks like a good candidate.

Reply:
I took a stab at a custom buffering solution but it quickly felt like reimplementing the paging logic the memory mapping gives us (just less performant) so I would stick with the memory mapped solution

regarding including a thirdparty in order to avoid bloat seems nice, but it is harder to modify (fix) and we are at the mercy of the owners for maintenance (in the library you linked the owner commented in one of the issues that they can't work on this project for now), since the current solution is less than 200 lines and essentially a wrapper around platform utils, and is tested on every application start, and is currently used in a single place, I feel like we should wait with replacing it with a thirdparty until we have multiple uses for it and/or more extensive requirements, like specifying offset/size

Reply:
I committed a reasonably fast solution without mmap. Most cases take <5ms on my machine, while the worst cases take around 40ms. I removed the log lines, because they would have required a much more complex solution, and it's probably not worth so much complexity. 
f263e07

Reply:
Does the :+1: mean that you agree with the changes and want to keep them as submitted? If there are any objections, we can continue to discuss here or on slack.

Reply:
yes, we can to keep it as submitted